### PR TITLE
[LV] Fix crash when vectorizing function calls with linear args.

### PR DESCRIPTION
**Pull request by labrinea**
llvm/lib/IR/Type.cpp:694:
    Assertion `isValidElementType(ElementType) && "Element type of a
    VectorType must be an integer, floating point, or pointer type."'
    failed.
Stack dump:
    llvm::FixedVectorType::get(llvm::Type*, unsigned int)
    llvm::VPWidenCallRecipe::execute(llvm::VPTransformState&)
    llvm::VPBasicBlock::execute(llvm::VPTransformState*)
    llvm::VPRegionBlock::execute(llvm::VPTransformState*)
    llvm::VPlan::execute(llvm::VPTransformState*)
    ...

Happens with function calls of void return type.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -504,7 +504,8 @@ void VPWidenCallRecipe::execute(VPTransformState &State) {
   for (unsigned Part = 0; Part < State.UF; ++Part) {
     SmallVector<Type *, 2> TysForDecl;
     // Add return type if intrinsic is overloaded on it.
-    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1)) {
+    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1) &&
```
[the remark this answers is not in the document]
Comment (labrinea) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Done

File: llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll [hunk 2]
```diff
@@ -0,0 +1,31 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call.*foo" --version 4
+; RUN: opt -passes=loop-vectorize,simplifycfg -force-vector-interleave=1 -force-vector-width=2 -S < %s | FileCheck %s
+
+; Make sure the vectorizer does not crash.
```
[the remark this answers is not in the document]
Comment (labrinea) [reply llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll:4]: Perhaps I misunderstoodm but originally in https://github.com/llvm/llvm-project/pull/76060#discussion_r1432823032 you suggested writing a target independent test.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -498,13 +498,15 @@ void VPWidenCallRecipe::execute(VPTransformState &State) {
          "DbgInfoIntrinsic should have been dropped during VPlan construction");
   State.setDebugLocFrom(CI.getDebugLoc());
 
+  Intrinsic::ID ID = VectorIntrinsicID;
```
Comment (mgabka) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:501]: Is it worth to create a local variable ? IMO it is better to use the "VectorIntrinsicID"

File: llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgabka) [reply llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll:4]: sorry, I think I mislead you, having a target independent test is not a bad idea, and I am not against.

 I think that having a full coverage in the AArch64 tests would be still good, as we can cover the scalable case, which can not be covered in the target independent tests, the _LLVM_ won't allow to add mappings for scalable vectors.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (labrinea) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:501]: My argument is that it helps with code indentation.

File: llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (labrinea) [reply llvm/test/Transforms/LoopVectorize/vector-call-linear-args.ll]: Done

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (labrinea) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Done

**Review by mgabka — APPROVED** [review]
Thanks for all the changes, LGTM.

**Pull request merged**